tools/lintapidiff.ml

```diff
@@ -0,0 +1,268 @@
+      match ty.pmty_desc with
+      | Pmty_signature lst -> List.fold_left self map lst
+      | Pmty_functor ({txt;_}, _, m) ->
+          let path = Path.Papply(path, Path.Pident (Ident.create txt)) in
+          add_module_type path m (inherits, map)
+      | _ -> map
+    in
+    let enter_path prefix path name ty attrs map =
+      let path = Path.Pdot (path, name.txt, 0) in
+      let inherits = f inherits name.loc attrs in
+      add_module_type path ty (inherits, map)
+    in
+    let add_module map m =
+      enter_path "module" path m.pmd_name m.pmd_type m.pmd_attributes map
+    in
+    match item.psig_desc with
+    | Psig_value vd ->
+        add_path ~f "value" path vd.pval_name vd.pval_attributes inherits map
+    | Psig_type (_,lst) ->
+        List.fold_left (fun map t ->
+            add_path ~f "type" path t.ptype_name t.ptype_attributes inherits map
+          ) map lst
+    | Psig_exception e ->
+        add_path ~f "exception" path e.pext_name e.pext_attributes inherits map
+    | Psig_module m -> add_module map m
+    | Psig_recmodule lst -> List.fold_left add_module map lst
+    | Psig_modtype s ->
+        begin match s.pmtd_type with
+        | None -> map
+        | Some ty ->
+            enter_path "module type" path s.pmtd_name ty s.pmtd_attributes map
+        end
+    | _ -> map
+
+  let add_items ~f path (inherits,map) items =
+    (* module doc *)
+    let inherits = List.fold_left (fun inherits -> function
+        | {psig_desc=Psig_attribute a}
+          when (Doc.get_doc ["ocaml.doc";"ocaml.text"][a] <> None) ->
+            f inherits (Location.none) [a]
+        | _ -> inherits
+      ) inherits items in
+    List.fold_left (add_item ~f path inherits) map items
+
+  let parse_file ~orig ~f ~init input =
+    try
+      let id =
+        orig |> Filename.chop_extension |> Filename.basename |>
+        String.capitalize_ascii |> Ident.create in
+      let ast = Pparse.file ~tool_name:Sys.argv.(0) Format.err_formatter input
+          Parse.interface Pparse.Signature in
+      Location.input_name := orig;
+      add_items ~f (Path.Pident id) (init,IdMap.empty) ast
+    with e ->
+      Format.eprintf "%a@." Location.report_exception e;
+      raise e
+end
+
+module Git = struct
+  let with_show ~f rev path =
+    let obj = rev ^ ":" ^ path in
+    let suffix = Printf.sprintf "-%s:%s" rev (Filename.basename path) in
+    let tmp = Filename.temp_file "" suffix in
+    let cmd = Printf.sprintf "git show %s >%s 2>/dev/null"
+        (Filename.quote obj) (Filename.quote tmp) in
+    Misc.try_finally (fun () ->
+        match Sys.command cmd with
+        | 0 -> Ok (f tmp)
+        | 128 -> Error `Not_found
+        | r ->
+            Location.errorf ~loc:(in_file obj) "exited with code %d" r |>
+            Format.eprintf "%a@." Location.report_error;
+            Error `Exit)
+      (fun () -> Misc.remove_file tmp)
+end
+
+module Diff = struct
+  type seen_info = {
+    not_seen: version option;
```

Comment:
Where is `not_seen` set? I understand from the report that it identifies a version in which the item is missing (I would thus find `option list` more natural), but it seems to be `None` everywhere in the code.